core/src/avm1/value.rs

```diff
@@ -671,51 +672,159 @@ fn f64_to_string(mut n: f64) -> Cow<'static, str> {
     }
 }
 
-/// Converts a `WStr` to an f64 based on the SWF version.
-fn string_to_f64(str: &WStr, swf_version: u8) -> f64 {
-    if swf_version < 5 {
-        use crate::avm1::globals::parse_float_impl;
-        let v = parse_float_impl(str.trim_start(), true);
-        if v.is_nan() {
-            return 0.0;
+/// Consumes an optional sign character.
+/// Returns whether a minus sign was consumed.
+fn parse_sign(s: &mut &WStr) -> bool {
+    if let Some(after_sign) = s.strip_prefix(b'-') {
+        *s = after_sign;
+        true
+    } else if let Some(after_sign) = s.strip_prefix(b'+') {
+        *s = after_sign;
+        false
+    } else {
+        false
+    }
+}
+
+/// Converts a `WStr` to an `f64`.
+///
+/// This function might fail for some invalid inputs, by returning `NaN`.
+///
+/// `strict` typically tells whether to behave like `Number()` or `parseFloat()`:
+/// * `strict == true` fails on trailing garbage (like `Number()`).
+/// * `strict == false` ignores trailing garbage (like `parseFloat()`).
+pub fn parse_float_impl(mut s: &WStr, strict: bool) -> f64 {
+    // Allow leading whitespace.
+    s = s.trim_start();
+
+    // Parse sign.
+    let is_negative = parse_sign(&mut s);
+    let after_sign = s;
+
+    // Validate digits before decimal point.
+    s = s.trim_start_matches(|c| (c as u8).is_ascii_digit());
+    let mut has_digits = s.len() < after_sign.len();
+    let mut exp = (after_sign.len() - s.len()) as i32 - 1;
+
+    // Validate digits after decimal point.
+    if let Some(after_dot) = s.strip_prefix(b'.') {
+        s = after_dot;
+        s = s.trim_start_matches(|c| (c as u8).is_ascii_digit());
+        has_digits |= s.len() < after_dot.len();
+    }
+
+    // Fail if we got no digits.
+    if !has_digits {
+        return f64::NAN;
+    }
+
+    // Handle exponent.
+    if let Some(after_e) = s.strip_prefix(b"eE".as_ref()) {
+        s = after_e;
+
+        // Parse exponent sign.
+        let exponent_is_negative = parse_sign(&mut s);
+
+        // Parse exponent itself.
+        let mut exponent: i32 = 0;
+        s = s.trim_start_matches(|c| match (c as u8 as char).to_digit(10) {
+            Some(digit) => {
+                exponent = exponent.wrapping_mul(10);
+                exponent = exponent.wrapping_add(digit as i32);
+                true
+            }
+            None => false,
+        });
+
+        // Apply exponent sign.
+        if exponent_is_negative {
+            exponent = exponent.wrapping_neg();
         }
-        return v;
+
+        exp = exp.wrapping_add(exponent);
     }
 
-    if str.is_empty() {
+    // Fail if we got digits, but we're in strict mode and not at end of string.
+    if strict && !s.is_empty() {
         return f64::NAN;
     }
 
+    // Finally, calculate the result.
+    let mut result = 0.0;
+    for c in after_sign {
+        if let Some(digit) = (c as u8 as char).to_digit(10) {
```

Comment:
The `as u8` casts ends up accepting characters >255, for example:

```
var s = String.fromCharCode(305);
trace(s);
trace(parseFloat(s)); // NaN in Flash, 1 in this PR
```

Reply:
Depending on how thorough you want to be in this PR, I wonder if it is worthwhile to make our own byte-to-digit fn that takes in a u16. Possibly we could use it for `parseInt` too.

Reply:
Good catch! I replaced the `as u8` conversions with `u8::try_from()`, which doesn't wrap around.